dm/syncer/syncer.go

```diff
@@ -2189,7 +2189,7 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 
 		// check pass SafeModeExitLoc and try to disable safe mode, but not in sharding or replacing error
 		safeModeExitLoc := s.checkpoint.SafeModeExitPoint()
-		if safeModeExitLoc != nil && shardingReSync == nil {
+		if s.safeMode.Enable() && safeModeExitLoc != nil && shardingReSync == nil {
```

Comment:
It's because `Syncer` dosen't enter safe-mode when `exitPoint` equal to `beginPosition` due to the safe_mode.go(L82-L92). So, when dm restart in this situation, dm will `SaveSafeModeExitPoint` and disable safe mode(cause error) in L2204 again without `s.safeMode.Enable()` of this line.

Therefore, I added it.

Reply:
in safe_mode.go line 84 we should save nil to `SafeModeExitPoint`

Reply:
However, I don't know what's any problems that checking safe-mode enable before disable safe-mode.🤔